Add mon rescheduling known issue info (#930)
replicatedhq/kurl.sh: src/markdown-pages/install-with-kurl/managing-nodes.md

This topic describes how to manage nodes on kURL clusters.
It includes procedures for how to safely reset, reboot, and remove nodes when performing maintenance tasks.

See the following sections:

* [ECKO Add-On Prerequisite](#ecko-add-on-prerequisite)
* [Reset a Node](#reset-a-node)
* [Reboot a Node](#reboot-a-node)
* [Remove a Node from Rook Ceph Clusters](#remove-a-node-from-rook-ceph-clusters)
   * [Rook Ceph and etcd Node Removal Requirements](#rook-ceph-and-etcd-node-removal-requirements)
   * [Rook Ceph Cluster Prerequisites](#rook-ceph-cluster-prerequisites)
   * [(Recommended) Manually Rebalance Ceph and Remove a Node](#recommended-manually-rebalance-ceph-and-remove-a-node)
   * [Remove Nodes with ECKO](#remove-nodes-with-ecko)

> * [Troubleshoot Node Removal](#troubleshoot-node-removal)

Reviewer: What about Known issues with Node Removal ?

Author: I'm going to leave it as Troubleshoot, since we are providing troubleshooting steps.

Complete the following prerequisites before you remove one or more nodes from a Rook Ceph cluster:

* Upgrade Rook Ceph to v1.4 or later.

   The two latest minor releases of Rook Ceph are actively maintained. It is recommended to upgrade to the latest stable release available. For more information, see [Release Cycle](https://rook.io/docs/rook/latest/Getting-Started/release-cycle/) in the Rook Ceph documentation.

>    Attempting to remove a node from a cluster that uses a Rook Ceph version earlier than v1.4 can cause Ceph to enter an unhealthy state. For example, see [Rook Ceph v1.0.4 is Unhealthy with Mon Pods Not Rescheduled](#rook-ceph-v104-is-unhealthy-with-mon-pods-not-rescheduled) under _Troubleshoot Node Removal_ below.

Reviewer: ```suggestion
   Attempting to remove a node from a cluster that uses a Rook Ceph version earlier than v1.4 might cause Ceph to enter an unhealthy state. For example, see [Rook Ceph v1.0.4 is Unhealthy with Mon Pods Not Rescheduled](#rook-ceph-v104-is-unhealthy-with-mon-pods-not-rescheduled) under _Troubleshoot Node Removal_ below.
```

I could reproduce the scenario but it also seems intermittent.

Complete the following prerequisites before you remove one or more nodes from a Rook Ceph cluster:

> * Upgrade Rook Ceph to v1.4 or later. Attempting to remove a node from a cluster that uses a Rook Ceph version earlier than v1.4 can cause Ceph to enter an unhealthy state. For example, see [Rook Ceph v1.0.4 is Unhealthy with Mon Pods Not Rescheduled](#rook-ceph-v104-is-unhealthy-with-mon-pods-not-rescheduled) under _Troubleshoot Node Removal_ below.

Reviewer: Could we also supplement here saying that ideally we recommend use and keep updated with the latest Rook version provided?

Reviewer: Also, wdyt about we link https://rook.io/docs/rook/latest/Getting-Started/release-cycle/ and say that most recent two minor Rook releases are those that are actively maintained by Rook itself.

Author: Great idea!

      ```
      kubectl -n rook-ceph exec deployment/rook-ceph-operator -- ceph status
      ```

>       **Note**: It is not recommended to use versions of Rook Ceph earlier than v1.4.0.

Reviewer: ```suggestion
      **Note**: It is no longer recommended to use versions of Rook Ceph earlier than v1.4.0.
```

Reviewer: @paigecalvert just a nit. Otherwise, all shows good for me.

Author: I took a look at this suggestion, and I recommend leaving this as "not required" rather than "no longer required" because the phrase "no longer" assumes that the user has knowledge about requirements from the past.

We try to keep the docs "timeless", without any reference to past or future. (Google has a good description of "timeless documentation" here, if you are interested: https://developers.google.com/style/timeless-documentation)

That said, please let me know if I misunderstood the purpose of this suggestion.

Reviewer: Thank you a lot for sharing that and help me indeed improve the wording skills !!!

   The EKCO purge script For information about the EKCO purge script, see [Purge Nodes](/docs/add-ons/ekco#purge-nodes) in _EKCO Add-on_.

1. Remove the node from the cluster.

## Troubleshoot Node Removal

This section includes information about troubleshooting issues with node removal in Rook Ceph clusters.

### Rook Ceph v1.0.4 is Unhealthy with Mon Pods Not Rescheduled

#### Symptom

When you run `kubectl -n rook-ceph exec deployment.apps/rook-ceph-operator -- ceph status` for a Rook Ceph v1.0.4 cluster, you see that Ceph is in an unhealthy state where failed mons are not rescheduled. 

For example:

> ```
> health: HEALTH_WARN
>         1/3 mons down, quorum a,c
> ```

Reviewer: ```suggestion
![unhelth-ceph-check](https://user-images.githubusercontent.com/7708031/214510937-e534f3e3-53e8-495f-93d1-3b23cda14c5d.png)
```

Reviewer: If you commit this suggestion we will have the image in the docs. WDYT?

> 

Reviewer: I think we need to add here the mon pod in Pending state as well;

![mon-pod-peding](https://user-images.githubusercontent.com/7708031/214511317-2da67544-65b8-4b0a-a4a5-14057f5b0271.png)

Author: Will do! Question: Will the value of the `app` field vary depending on the name of the application? `rook-ceph-mon` is just an example app that you've creating for testing, correct?

(I'll update the command to make it clear that this is a placeholder value if so)

Author: Done!

#### Cause

> This is caused by a known issue in Rook Ceph v1.0.4 where failed mons are not removed during mon failover. For more information, see [Mon is never rescheduled](https://github.com/rook/rook/issues/2262#issuecomment-460898915) in the rook GitHub repository.

Reviewer: ```suggestion
This is caused by a known issue in Rook Ceph v1.0.4 where the rook-ceph-mon-endpoints still have mapped the node that was purged. For more information, see [Mon is never rescheduled](https://github.com/rook/rook/issues/2262#issuecomment-460898915) in the rook GitHub repository.
```

Author: Thank you--updated.

#### Solution

> To resolve this issue, manually stop the Rook Ceph operator, remove the failed mon from the configmap, delete the failed mon, and then rescale the operator.  

Reviewer: What about we add here the steps with the commands as we have in the community docs ? 
```suggestion
To resolve this issue you will need to perform the following steps: 
- Stop the Rook Operator by running: kubectl -n rook-ceph scale --replicas=0 deployment.apps/rook-ceph-operator
- Edit the configmap rook-ceph-mon-endpoints to (carefully) remove the failed mon from the list with the command kubect -n rook-ceph edit configmaps rook-ceph-mon-endpoints. You will see that this config map still with data info from the node that was purged. 
> ATTENTION: If you remove the wrong data endpoint it might cause unexpected behaviours as cause data loss.
- Delete the mon pod in Pending state. You can found its name by running kubectl -n rook-ceph get pod -l app=rook-ceph-mon and you can remove by running kubectl -n rook-ceph delete pod <mon-pod-name-in-pending-state>
- Then, rescale the Rook Operator by running: kubectl -n rook-ceph scale --replicas=1 deployment.apps/rook-ceph-operator
```

Author: Sounds good to me. Updated 👍

> 

Reviewer: Please, ensure that after the changes made above all mon pods will be running and the Ceph status is healthy.

Reviewer: I think we need to add the step to verify that the fix was done right.

Author: I added two steps to the end of the procedure to check that the mon pods are all running and to check that ceph is healthy

   ```
   kubect -n rook-ceph edit configmaps rook-ceph-mon-endpoints
   ```

  _**Warning**_: Ensure that you remove the correct rook-ceph-mon-endpoint from the `mapping` field in the ConfigMap. Removing the wrong rook-ceph-mon-endpoint can cause unexpected behavior, including data loss.

1. Find the name of the Pending mon pod:

   ```
   kubectl -n rook-ceph get pod -l app=rook-ceph-mon
   ```

1. Delete the Pending mon pod:

   ```
   kubectl -n rook-ceph delete pod MON_POD_NAME
   ```
   Replace `MON_POD_NAME` with the name of the mon pod that is in a Pending state from the previous step. 

1. Rescale the operator:

   ```
   kubectl -n rook-ceph scale --replicas=1 deployment.apps/rook-ceph-operator
   ```

1. Verify that all mon pods are running:

   ```
   kubectl -n rook-ceph get pod -l app=rook-ceph-mon
   ```
   The output of this command shows that each mon pod has a `Status` of `Running`.

1. Verify that Ceph is in a healthy state:

   ```
   kubectl -n rook-ceph exec deployment.apps/rook-ceph-operator -- ceph status
   ```

   The output of this command shows `health: HEALTH_OK` if Ceph is in a healthy state.

> 1. After confirming that Ceph is in a healthy state, upgrade Rook Ceph to v1.4 or later before attempting to managing nodes in the cluster.

Reviewer 2: ```suggestion
1. After confirming that Ceph is in a healthy state, upgrade Rook Ceph to v1.4 or later before attempting to manage nodes in the cluster.
```

Author: Good catch, thank you

   The EKCO purge script For information about the EKCO purge script, see [Purge Nodes](/docs/add-ons/ekco#purge-nodes) in _EKCO Add-on_.

1. Remove the node from the cluster.

## Troubleshoot Node Removal

This section includes information about troubleshooting issues with node removal in Rook Ceph clusters.

### Rook Ceph v1.0.4 is Unhealthy with Mon Pod Pending

#### Symptom

After you remove a node from a Rook Ceph v1.0.4 cluster and you run `kubectl -n rook-ceph exec deployment.apps/rook-ceph-operator -- ceph status`, you see that Ceph is in an unhealthy state where a Ceph monitor (mon) is down. 

For example:

```
health: HEALTH_WARN
        1/3 mons down, quorum a,c
```

When you run `kubectl -n rook-ceph get pod -l app=rook-ceph-mon`, you see that the mon pod is in a Pending state. 

For example:

```
NAME             READY  STATUS   RESTARTS  AGE
rook-ceph-mon-a  1/1    Running  0         20m
rook-ceph-mon-b  0/1    Pending  0         9m45s
rook-ceph-mon-c  1/1    Running  0         13m
```

#### Cause

> This is caused by a known issue in Rook Ceph v1.0.4 where the rook-ceph-mon-endpoints ConfigMap still maps a node that was removed. For more information, see [Mon is never rescheduled](https://github.com/rook/rook/issues/2262#issuecomment-460898915) in the rook GitHub repository.

Reviewer: ```suggestion
This known issue in Rook Ceph v1.0.4 shows to be caused by the rook-ceph-mon-endpoints ConfigMap still maps a node that was removed which seems similar to the issue [Mon is never rescheduled](https://github.com/rook/rook/issues/2262#issuecomment-460898915) in the rook GitHub repository.
```

> #### Solution

Reviewer: @paigecalvert we might want to write something here to let users know that it is a workaround to make the Ceph came back to health state in order to upgrade. However,  the best recommended solution is upgrade and no longer use this version.

> #### Solution

Reviewer: ```suggestion
#### How to sort out the Ceph warning state to upgrade
```

Reviewer: I think is better we avoid the word solution and goes in the direction of that is an workaround to came back to health state to upgrade which is the best option.

Reviewer: Our solution, I think, is more to upgrade. 
ihmo would be better if we could avoid letting people think that is fine and that they should do that to solve the issue and continue to use this version. This is more like workarounds to return to the Ceph for a healthy state. It is not an official and tested operation for all possible edge-case scenarios. It is not either doc in the Rook docs.

Author: "Our solution, I think, is more to upgrade. ihmo would be better if we could avoid letting people think that is fine and that they should do that to solve the issue and continue to use this version" > The content that I wrote does tell people that the solution is that they must upgrade. To help reinforce this idea, I changed the title to "Workaround".

> To resolve this issue, you must manually delete the mapping to the removed node from the rook-ceph-mon-endpoints ConfigMap.

Reviewer: ```suggestion
To workaround this issue and try to recover the Ceph health state, you might able to do it by manually editing the mapping to the removed the info about node that was purged from the rook-ceph-mon-endpoints ConfigMap.
```

> For more information about these steps, see [Managing nodes when the previous Rook version is in use might leave Ceph in an unhealthy state where mon pods are not rescheduled](https://community.replicated.com/t/managing-nodes-when-the-previous-rook-version-is-in-use-might-leave-ceph-in-an-unhealthy-state-where-mon-pods-are-not-rescheduled/1099/1) in _Replicated Community_.

Reviewer: Should it still here ? 
If so, would not be better add at the bottom?

To delete the mapping for the removed node:

1. Stop the Rook Ceph operator:

   ```
   kubectl -n rook-ceph scale --replicas=0 deployment.apps/rook-ceph-operator
   ```

1. Edit the rook-ceph-mon-endpoints ConfigMap to delete the mapping to the removed node:

   ```
   kubect -n rook-ceph edit configmaps rook-ceph-mon-endpoints
   ```

>   _**Warning**_: Ensure that you remove the correct rook-ceph-mon-endpoint from the list. Removing the wrong rook-ceph-mon-endpoint can cause data loss.

Reviewer: ```suggestion
  _**Warning**_: Ensure that you remove the correct rook-ceph-mon-endpoint from the list. Removing the wrong rook-ceph-mon-endpoint can cause unexpected behaviours.
```

We do not know what might can happens if someone edit the wrong value by mistake. 
So, I think is important we highlight that but it might not be good we do any affirmation about what would happens

   The EKCO purge script For information about the EKCO purge script, see [Purge Nodes](/docs/add-ons/ekco#purge-nodes) in _EKCO Add-on_.

1. Remove the node from the cluster.

## Troubleshoot Node Removal

This section includes information about troubleshooting issues with node removal in Rook Ceph clusters.

### Rook Ceph v1.0.4 is Unhealthy with Mon Pod Pending

#### Symptom

> After you remove a node from a Rook Ceph v1.0.4 cluster and you run `kubectl -n rook-ceph exec deployment.apps/rook-ceph-operator -- ceph status`, you see that Ceph is in an unhealthy state where a Ceph monitor (mon) is down. 

Reviewer: ```suggestion
After you remove a node from a Rook Ceph v1.0.4 cluster and you run `kubectl -n rook-ceph exec deployment.apps/rook-ceph-operator -- ceph status`, you see that Ceph is in an unhealthy state where a Ceph monitor (mon) is down and under the services that we have one or more out of quorum.
```

For example:

```
health: HEALTH_WARN

>         1/3 mons down, quorum a,c

Reviewer: @paigecalvert can we add the part of that the service is out of quorum as well ?

   ```
   kubect -n rook-ceph edit configmaps rook-ceph-mon-endpoints
   ```

  _**Warning**_: Ensure that you remove the correct rook-ceph-mon-endpoint from the `mapping` field in the ConfigMap. Removing the wrong rook-ceph-mon-endpoint can cause unexpected behavior, including data loss.

1. Find the name of the Pending mon pod:

   ```
   kubectl -n rook-ceph get pod -l app=rook-ceph-mon
   ```

1. Delete the Pending mon pod:

   ```
   kubectl -n rook-ceph delete pod MON_POD_NAME
   ```
   Replace `MON_POD_NAME` with the name of the mon pod that is in a Pending state from the previous step. 

1. Rescale the operator:

   ```
   kubectl -n rook-ceph scale --replicas=1 deployment.apps/rook-ceph-operator
   ```

1. Verify that all mon pods are running:

   ```
   kubectl -n rook-ceph get pod -l app=rook-ceph-mon
   ```
   The output of this command shows that each mon pod has a `Status` of `Running`.

1. Verify that Ceph is in a healthy state:

   ```
   kubectl -n rook-ceph exec deployment.apps/rook-ceph-operator -- ceph status
   ```

   The output of this command shows `health: HEALTH_OK` if Ceph is in a healthy state.

> 1. After confirming that Ceph is in a healthy state, upgrade Rook Ceph to v1.4 or later before attempting to managing nodes in the cluster.

Author: I added in this last step to explain that they really need to upgrade in order to address this issue.

Reviewer: Should we say check if you can upgrade the Rook version used?
Not sure if we can say just upgrade and then if they are using a vendor app that does not work with the upper version?

Author: We state elsewhere in the kURL documentation that our requirement is to use v1.4 or later, so this should not be a concern.
